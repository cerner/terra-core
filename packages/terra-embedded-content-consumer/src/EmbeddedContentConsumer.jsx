import React from 'react';
import PropTypes from 'prop-types';
import { Consumer } from 'xfc';

const propTypes = {
  /**
   * The source URL of the content to load.
   */
  src: PropTypes.string.isRequired,
  /**
   * Notifies the component that the container has been mounted. Provides a reference
   * to this component to allow triggering messages on the embedded application.
   */
  onMount: PropTypes.func,
  /**
   * Notifies the component that the container has been launched.
   */
  onLaunch: PropTypes.func,
  /**
   * Notifies the component that the container has been authorized.
   */
  onAuthorize: PropTypes.func,
  /**
   * The component can be configured with consumer frame options.
   * See xfc consumer configuration for details: https://github.com/cerner/xfc
   */
  options: PropTypes.object,
  /**
  * A set of event handlers keyed by the event name.
  * Note: Binding the event handler is necessary to make `this` work in the callback.
  */
  eventHandlers: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    handler: PropTypes.func,
  })),
};

class EmbeddedContentConsumer extends React.Component {
  componentDidMount() {
    // Mount the provided source as the application into the content wrapper.
    this.xfcFrame = Consumer.mount(this.embeddedContentWrapper, this.props.src, this.props.options);

    // Notify that the consumer frame has mounted.
    if (this.props.onMount) {
      this.props.onMount(this.xfcFrame);
    }

    // Attach the event handlers to the xfc frame.
    this.addEventListener('xfc.launched', this.props.onLaunch);
    this.addEventListener('xfc.authorized', this.props.onAuthorize);

    // Attach custom event handlers to the xfc frame.
    this.addEventListeners(this.props.eventHandlers);
  }

  addEventListener(eventName, eventHandler) {
    if (eventName && eventHandler) {
      this.xfcFrame.on(eventName, eventHandler);
    }
  }

  addEventListeners(customEvents) {
    (customEvents || []).forEach(e => this.addEventListener(e.key, e.handler));
  }

  render() {
    const {
      src,
      onMount,
      onLaunch,
      onAuthorize,
      options,
      eventHandlers,
      ...customProps
    } = this.props;

    return (
      <div
        {...customProps}
        ref={(element) => { this.embeddedContentWrapper = element; }}
      />
    );
  }
}

EmbeddedContentConsumer.propTypes = propTypes;

export default EmbeddedContentConsumer;
