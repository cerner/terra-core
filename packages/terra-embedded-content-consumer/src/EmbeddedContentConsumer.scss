/* stylelint-disable max-nesting-depth */
/* stylelint-disable selector-class-pattern */
/* stylelint-disable selector-max-compound-selectors */
:local {
  .fill {
    height: 100%;

    :global {
      .xfc {
        height: 100%;
     
        iframe {
          height: 100%;
          width: 100%;
        }
      }
    }
  }
}
