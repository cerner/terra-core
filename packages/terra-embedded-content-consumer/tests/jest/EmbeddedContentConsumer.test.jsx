import React from 'react';
import { Consumer } from 'xfc';
import EmbeddedContentConsumer from '../../lib/EmbeddedContentConsumer';

// Snapshot Tests
it('should render the embedded content consumer container', () => {
  const embeddedContentConsumer = (
    <EmbeddedContentConsumer
      src="https://www.google.com/"
    />
  );

  const wrapper = render(embeddedContentConsumer);
  expect(wrapper).toMatchSnapshot();
});

it('should render the embedded content consumer with custom class names', () => {
  const embeddedContentConsumer = (
    <EmbeddedContentConsumer
      src="https://www.google.com/"
      className="container"
    />
  );

  const wrapper = render(embeddedContentConsumer);
  expect(wrapper).toMatchSnapshot();
});

it('should validate the inputs', () => {
  const src = 'https://www.google.com/';
  let frame;
  const onMount = (xfcFrame) => { frame = xfcFrame; };
  const onLaunch = () => {};
  const onAuthorize = () => {};
  const resizeConfig = { scrolling: false, fixedHeight: '100%', fixedWidth: '100%' };
  const options = { secret: 'SecretKey', resizeConfig };
  const customEvents = [{ key: 'eventKey', handler: () => {} }];

  const embeddedContentConsumer = (
    <EmbeddedContentConsumer
      src={src}
      onMount={onMount}
      onLaunch={onLaunch}
      onAuthorize={onAuthorize}
      options={options}
      eventHandlers={customEvents}
    />
  );

  Consumer.init();
  const wrapper = mount(embeddedContentConsumer);

  expect(frame).toBeTruthy();
  expect(wrapper.instance().props.src).toBe(src);
  expect(wrapper.instance().props.onMount).toBe(onMount);
  expect(wrapper.instance().props.onLaunch).toBe(onLaunch);
  expect(wrapper.instance().props.onAuthorize).toBe(onAuthorize);
  expect(wrapper.instance().props.options).toBe(options);
  expect(wrapper.instance().props.options.resizeConfig).toBe(resizeConfig);
  expect(wrapper.instance().props.eventHandlers).toBe(customEvents);
});
