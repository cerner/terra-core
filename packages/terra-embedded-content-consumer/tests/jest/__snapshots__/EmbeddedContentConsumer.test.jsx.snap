// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the embedded content consumer container 1`] = `<div />`;

exports[`should render the embedded content consumer with custom class names 1`] = `
<div
  class="container"
/>
`;
